src/unsync/mpsc.rs

```diff
@@ -0,0 +1,127 @@
+//! A multi-producer, single-consumer, futures-aware, FIFO queue with back
+//! pressure, for use communicating between tasks on the same thread.
+
+use std::collections::VecDeque;
+use std::rc::Rc;
+use std::cell::RefCell;
+
+use task::{self, Task};
+use sink::SendError;
+use {Async, AsyncSink, Poll, StartSend, Sink, Stream};
+
+/// Creates a bounded in-memory channel with pre-allocated storage.
+///
+/// This method creates concrete implementations of the `Stream` and `Sink`
+/// traits which can be used to communicate a stream of values between tasks
+/// with backpressure. The channel capacity is exactly `buffer`.
+pub fn channel<T>(buffer: usize) -> (Sender<T>, Receiver<T>) {
+    let shared = Rc::new(RefCell::new(Shared {
+        buffer: VecDeque::with_capacity(buffer),
```

Comment:
Perhaps the capacity could be stored on the side and this could just allocate a `VecDeque`? We may not actually hit the capacity ever and this way we'd save memory for large `buffer` sizes passed in (perhaps)